### PR TITLE
Parameter filtering

### PR DESCRIPTION
**Pull request by SchrodingersGat**
- Closes https://github.com/inventree/InvenTree/issues/3080
- Closes https://github.com/inventree/InvenTree/issues/4209

This PR adds a number of key improvements to the concept of "part parameterrs"

### TODO

- [x] ~~Allows different "types" of parameter templates to be defined~~ (will implement later)
- [x] ~~Validates entered values against specified template type~~ (will implement later)
- [x] Add physical units to parameters - https://github.com/inventree/InvenTree/issues/4209
- [x] Automatically convert to base units based on SI units
- [x] Calculates and stores numerical value of parameter value (for efficient sorting)
- [x] Data migration to calculate "numerical value" for existing parameters
- [x] Unit test for above data migration
- [x] Background task to re-calculate parameter values if a template changes
- [x] Performs sorting of parts by parameter value in table
- [ ] ~~Performs querying of parts by parameter value in table (e.g. `eq` / `lt` / `gt` )~~ - moved to https://github.com/inventree/InvenTree/issues/4851
- [x] Documentation
- [x] Updates to demo dataset with new physical parameters - https://github.com/inventree/demo-dataset/pull/45

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
@matmair should we migrate the existing "units" field (which is just a string, no functional value) to a physical unit?

File: InvenTree/part/admin.py [hunk 1]
```diff
@@ -122,9 +122,9 @@ class Meta(PartResource.Meta):
         ]
 
 
-class StocktakeInline(admin.TabularInline):
-    """Inline for part stocktake data"""
-    model = models.PartStocktake
```
Comment (matmair) [inline InvenTree/part/admin.py:127]: Question: Is there no need for the old inline anymore?

File: InvenTree/part/models.py [hunk 2]
```diff
@@ -3295,6 +3296,9 @@ class PartParameterTemplate(MetadataMixin, models.Model):
     Attributes:
         name: The name (key) of the Parameter [string]
         units: The units of the Parameter [string]
+        description: Description of the parameter [string]
+        type: The data type of the Parameter [integer]
+        choices: A comma-separated list of choices for the Parameter [string]
```
Comment (matmair) [inline InvenTree/part/models.py:3301]: Question: Where are type / choices defined?

**Review by matmair — COMMENTED** [review]
Very cool development;

Some ideas/discussion points:
- I think it could be beneficial to be able to deactivate all calculations on a field
- It could be interesting to mark a part parameter as locked. Ie. the user can not change the parameter value - only plugins can
- An API endpoint or query string that returns a parameter in a specific unit would be nice
- This PR assumes we use SI; That should be documented somewhere

**Comment by matmair** [comment]
> @matmair should we migrate the existing "units" field (which is just a string, no functional value) to a physical unit?
> 
> Edit: I've started to dive into using [pint](https://pypi.org/project/Pint/) and looks like I have a good path forward for dev

We can certainly try; not sure if this would lead to errors in non-English languages. Maybe we could add a button in the settings and prompt all superusers?

File: InvenTree/part/admin.py (hunk 1 above)
Comment (SchrodingersGat) [reply InvenTree/part/admin.py:127]: IMO no, it has been outdated since the recent stocktake updates

File: InvenTree/part/models.py (hunk 2 above)
Comment (SchrodingersGat) [reply InvenTree/part/models.py:3301]: @matmair I'm going to move this into https://github.com/inventree/InvenTree/issues/4835

File: InvenTree/build/migrations/0013_auto_20200425_0507.py [hunk 3]
```diff
@@ -53,5 +49,5 @@ class Migration(migrations.Migration):
             field=models.PositiveIntegerField(db_index=True, default=0, editable=False),
             preserve_default=False,
         ),
-        migrations.RunPython(update_tree, reverse_code=nupdate_tree),
+        migrations.RunPython(update_tree, reverse_code=migrations.RunPython.noop),
```
Comment (matmair) [inline InvenTree/build/migrations/0013_auto_20200425_0507.py:52]: Comment: Nice, I did not know that

**Review by matmair — APPROVED** [review]

File: InvenTree/build/migrations/0013_auto_20200425_0507.py (hunk 3 above)
Comment (SchrodingersGat) [reply InvenTree/build/migrations/0013_auto_20200425_0507.py:52]: New to me also :)

**Comment by eeintech** [comment]
@SchrodingersGat Is it supposed to work on the demo version (engineer user)? I am seeing a few problems:

* Parameters tabs did not filter parts by the category selected
*  Only part name shows as there is no option to add parameters in new columns
* 
image

* Only common filters are available, no parameters in the list

image

I am not sure where to look for this new feature if not there?

**Comment by SchrodingersGat** [comment]
@eeintech you might have tried the demo server before the new version was loaded:

image

> Parameters tabs did not filter parts by the category selected

This looks like a bug, I'll look into it

**Comment by second-string** [comment]
I'm a bit confused, and maybe just misunderstanding, but I still cannot properly filter on parameter values. Both on my own server, and on the demo server, if I search for text that's in one of the parameter columns (but NOT in the part name/description), then those parts don't come up.

For example, on the demo server if you search for 'canvas' on the Parameters page, only one result comes up that has 'canvas' in the name. However there are multiple parts that have canvas as a parameter value. Clicking the sort arrow on a column properly sorts, but I can't figure out how to filter down by specific  parameter values.

There are also no params available in the filter dropdown, as eeintech showed in their screenshot above.

**Comment by SchrodingersGat** [comment]
@second-string the parametric table can currently be "sorted" by parameter value, but not filtered.

With regards to searching by parameter value - this is a good addition but currently the search is only against the parts, not the parameter values. I'll raise a new issue to look into that.

**Pull request merged**